src/cargo/sources/path.rs

```diff
@@ -834,10 +854,27 @@ fn list_files_walk(
 
             true
         });
+
+    let mut current_symlink_dir = None;
     for entry in walkdir {
         match entry {
             Ok(entry) => {
                 let file_type = entry.file_type();
+
+                match current_symlink_dir.as_ref() {
+                    Some(dir) if entry.path().starts_with(dir) => {
+                        // Still walk under the same parent symlink dir, so keep it
+                    }
+                    Some(_) | None => {
+                        // Not under any parent symlink dir, update the current one.
+                        current_symlink_dir = if file_type.is_dir() {
+                            entry.path_is_symlink().then(|| entry.path().to_path_buf())
+                        } else {
+                            None
+                        };
+                    }
+                }
+
```

Comment:
This assumes child files are walked before child directories, is that valid?

- `a/`
  - `b/`
  - `c/`
  - `d.txt`

Walking that:
1. `a/`, `current_symlink_dir='a'`
2. `a/b/`, `current_symlink_dir='a/b'`
3. `a/c/`, `current_symlink_dir='a/c'`
4. `a/d.txt`, `current_symlink_dir=None`

Reply:
Huh, we may need a stack.

Reply:
Probably not?

- `a/`
  - `b/`
  - `c/`
  - `d.txt`

Walking that:
1. `a/`, `current_symlink_dir='a'`
2. `a/b/`, `current_symlink_dir='a'` because still `.starts_with(dir)`
3. `a/c/`, `current_symlink_dir='a'` because still `.starts_with(dir)`
4. `a/d.txt`, `current_symlink_dir='a'` because still `.starts_with(dir)`

We keep the farthest symlink directory as long as possible, until we walk to the other completely outside the previous symlink directory.

Reply:
Given this project layout, I got these from `list_files`:

```
./
├── dir/
│   └── b
├── src/
│   ├── dir -> ../dir/
│   ├── regular-dir/
│   │   ├── symlink-dir -> ../../dir/
│   │   └── c
│   ├── symlink-dir -> ../dir/
│   ├── a
│   ├── lib.rs
│   └── z
├── symlink-src -> src/
├── symlink-src-but-actually-not/
│   └── x
└── Cargo.toml
```

```rust
[src/cargo/ops/cargo_package/vcs.rs:49:5] src_files = [
    PathEntry {
        path: "/foo/Cargo.toml",
        ty: File,
        under_symlink_dir: false,
    },
    PathEntry {
        path: "/foo/symlink-src/regular-dir/symlink-dir/b",
        ty: File,
        under_symlink_dir: true,
    },
    PathEntry {
        path: "/foo/symlink-src/regular-dir/c",
        ty: File,
        under_symlink_dir: true,
    },
    PathEntry {
        path: "/foo/symlink-src/symlink-dir/b",
        ty: File,
        under_symlink_dir: true,
    },
    PathEntry {
        path: "/foo/symlink-src/lib.rs",
        ty: File,
        under_symlink_dir: true,
    },
    PathEntry {
        path: "/foo/symlink-src/z",
        ty: File,
        under_symlink_dir: true,
    },
    PathEntry {
        path: "/foo/symlink-src/a",
        ty: File,
        under_symlink_dir: true,
    },
    PathEntry {
        path: "/foo/symlink-src/dir/b",
        ty: File,
        under_symlink_dir: true,
    },
    PathEntry {
        path: "/foo/.gitignore",
        ty: File,
        under_symlink_dir: false,
    },
    PathEntry {
        path: "/foo/dir/b",
        ty: File,
        under_symlink_dir: false,
    },
    PathEntry {
        path: "/foo/symlink-src-but-actually-not/x",
        ty: File,
        under_symlink_dir: false,
    },
    PathEntry {
        path: "/foo/src/regular-dir/symlink-dir/b",
        ty: File,
        under_symlink_dir: true,
    },
    PathEntry {
        path: "/foo/src/regular-dir/c",
        ty: File,
        under_symlink_dir: false,
    },
    PathEntry {
        path: "/foo/src/symlink-dir/b",
        ty: File,
        under_symlink_dir: true,
    },
    PathEntry {
        path: "/foo/src/lib.rs",
        ty: File,
        under_symlink_dir: false,
    },
    PathEntry {
        path: "/foo/src/z",
        ty: File,
        under_symlink_dir: false,
    },
    PathEntry {
        path: "/foo/src/a",
        ty: File,
        under_symlink_dir: false,
    },
    PathEntry {
        path: "/foo/src/dir/b",
        ty: File,
        under_symlink_dir: true,
    },
]
```

Reply:
Yeah, I forgot you only track the root-most one.  So this is likely fine.  Still feels confusing or brittle but likely it will be fine.